File: Leetcode/12.整数转罗马数字.cpp
/*
 * @lc app=leetcode.cn id=12 lang=cpp
 *
 * [12] 整数转罗马数字
 */
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/*2021-09-27-20-50*/

// from https://github.com/fengwei2002/Algorithm
/*
 ! 找规律
 !
 *
 * 优雅的解决问题
 https://leetcode-cn.com/problems/integer-to-roman/solution/lc12-fengwei2002-by-fengwei2002-r2an/
 */

// @lc code=start
class Solution {
   public:
    string intToRoman(int num) {
        int values[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        string reps[] = {
            "M",  "CM", "D",  "CD", "C",  "XC", "L",
            "XL", "X",  "IX", "V",  "IV", "I",
        };

        string res;
        for (int i = 0; i < 13; i++) {
            while (num >= values[i]) {
                num -= values[i];
                res += reps[i];
            }
        }

        return res;
    }
};
// @lc code=end

File: Leetcode/12.整数转罗马数字.go
/*
 * @lc app=leetcode.cn id=12 lang=golang
 *
 * [12] 整数转罗马数字
 */
package leetcode

/* --- 2022-04-06-20-01 --- */
// https://github.com/fengwei2002/Algorithm

// @lc code=start

type mapping struct {
	Value  int
	Symbol string
}

var hash []mapping = []mapping{
	{1000, "M"},
	{900, "CM"},
	{500, "D"},
	{400, "CD"},
	{100, "C"},
	{90, "XC"},
	{50, "L"},
	{40, "XL"},
	{10, "X"},
	{9, "IX"},
	{5, "V"},
	{4, "IV"},
	{1, "I"},
}

func intToRoman(num int) string {
	roman := make([]byte, 0)
	for _ /*index*/, v := range hash {
		for num >= v.Value {
			num -= v.Value
			roman = append(roman, []byte(v.Symbol)...)
		}
	}
	return string(roman)
}

// @lc code=end
